nuxeo-runtime/nuxeo-runtime-migration/src/main/java/org/nuxeo/runtime/migration/MigrationServiceImpl.java

```diff
@@ -0,0 +1,286 @@
+        String toState = atomic(id, kv -> {
+            String state = kv.getString(id);
+            String currentStep = kv.getString(id + STEP);
+            if (state == null && currentStep == null) {
+                // migration never started, assumed at first state
+                state = descr.states.keySet().iterator().next(); // LinkedHashMap guarantees order
+            } else if (state == null) {
+                if (currentStep.equals(step)) {
+                    // already running
+                    return null;
+                }
+                throw new IllegalArgumentException("Migration: " + id + " already running step: " + currentStep);
+            }
+            MigrationStateDescriptor stateDescr = descr.states.get(state);
+            if (stateDescr == null) {
+                throw new RuntimeException("Invalid current state: " + state + " for migration: " + id);
+            }
+            if (!stepDescr.fromState.equals(state)) {
+                throw new IllegalArgumentException(
+                        "Invalid step: " + step + " for migration: " + id + " in state: " + state);
+            }
+            String time = String.valueOf(System.currentTimeMillis());
+            kv.put(id + STEP, step);
+            kv.put(id + START_TIME, time);
+            kv.put(id + PING_TIME, time);
+            kv.put(id + PROGRESS_MESSAGE, "");
+            kv.put(id + PROGRESS_NUM, "0");
+            kv.put(id + PROGRESS_TOTAL, "-1");
+            kv.put(id, (String) null);
+            return stepDescr.toState;
+        });
+        if (toState == null) {
+            // already running
+            return;
+        }
+
+        KeyValueStore keyValueStore = getKeyValueStore();
+        MigrationContext migrationContext = new MigrationContext() {
+            @Override
+            public void reportProgress(String message, long num, long total) {
+                keyValueStore.put(id + PROGRESS_MESSAGE, message);
+                keyValueStore.put(id + PROGRESS_NUM, String.valueOf(num));
+                keyValueStore.put(id + PROGRESS_TOTAL, String.valueOf(total));
+            }
+        };
+        Runnable runnable = () -> {
+            migrator.run(migrationContext);
+            atomic(id, kv -> {
+                kv.put(id, toState);
+                kv.put(id + STEP, (String) null);
+                kv.put(id + START_TIME, (String) null);
+                kv.put(id + PING_TIME, (String) null);
+                kv.put(id + PROGRESS_MESSAGE, (String) null);
+                kv.put(id + PROGRESS_NUM, (String) null);
+                kv.put(id + PROGRESS_TOTAL, (String) null);
+            });
+        };
+
+        runnable.run(); // TODO XXX in a thread
+    }
+
+    /**
+     * Execute while setting a lock
+     */
+    protected <T> T atomic(String id, Function<KeyValueStore, T> function) {
+        KeyValueStore kv = getKeyValueStore();
+        String lock = String.valueOf(System.currentTimeMillis());
```

Comment:
may be rename "lock" to "value" here, could be nice to have the node name + timestamp for debug purpose